# Week2_2_Data Management_1

In this lab, we begin with the introduction of the Pandas for data management. You will be manipulating a new type of data called *DataFrame* which is the spreadsheet in Python. The dataset for this lab comes from a resale housing sample in Beijing from 2012 to 2016. 

Specifically. we will cover the following contents in this lab:


<p><a id="table"> </a></p>
<h3 id="Table-of-Contents">Table of Contents<a class="anchor-link" href="#Table-of-Contents">¶</a></h3>
<ol>
<h3>1 <a href="#pythonpackages">Python Packages</a></h3>
</ol>
<ol>
<h3>2 <a href="#pandasbasics">Pandas basics</a></h3>
2.1 <a href="#seriesanddataFrame">Series and DataFrame</a><br/>
2.2 <a href="#importandexportdata">Import and export data</a><br/>  
2.3 <a href="#exploringourdata">Exploring our data</a><br/>
2.4 <a href="#indexingandslicingofdataFrame">Indexing and Slicing of DataFrame</a><br/>   
2.5 <a href="#filtering">Filtering</a><br/>
2.6 <a href="#modifyingelements">Modifying elements</a><br/>
2.7 <a href="#pandasarithmetics">Pandas Arithmetics</a><br/>
2.8 <a href="#groupby">Groupby function</a><br/>    
</ol>



<p><a id="Python packages"> </a></p>
<h2 id="1-Python-packages">1 <a href="#table">Python packages</a><a class="anchor-link" href="#1-python-packages">¶</a></h2>

**Built-in Packages (or called The Python Standard Library)**: Anaconda comes with many Python built-in packages (also referred to as *modules* or *libraries*) that offer ready-to-use solutions to common programming problems. These packages have already been installed with Anaconda, and can be accessed using the *import* keyword. 

For today's lab, Pandas can be imported by typing `import pandas as pd` in your current script.

Here we load this package by `import` and give it a name "pd" using `as`, so that we do not need to type "pandas" each time we use it.

In [6]:
# import pandas and named it "pd" in the current python script
import pandas as pd

**Other Packages**: 
- Some packages are not installed with Anaconda, and We need to *install* them before importing to current script. For example, geopandas.

- **A Python virtual environment** is a tool for dependency management and project isolation. It allows packages to be installed locally in an isolated directory for a particular project, as opposed to being installed globally (i.e. as part of a system-wide Python).


<p><a id="pandasbasics"> </a></p>
<h2 id="2pandasbasics">2 <a href="#table">Pandas basics</a><a class="anchor-link" href="#2Pandasbasics">¶</a></h2>

- [Pandas](http://pandas.pydata.org/) is a widely used Python library for data analysis. 

### Easy-to-use data structures
- In pandas, the data is typically stored in a data structure called a DataFrame that looks like a typical table with rows and columns (+indices and column names), where columns can contain data of different data types. Thus, it is similar in some sense to how data is stored in Excel. Using this structure, we can use operations like arithmetic, columns and rows selection, columns and rows addition etc.


### Combines functionalities from many Python modules
- Pandas is established based on another package numpy&mdash;which supports the multi-dimensional arrays and the mathematical functions for scientific computing. It is much more than an easier-to-use Numpy as it also combines many functionalities from other Python libraries such as [matplotlib (plotting)](https://matplotlib.org/) and [scipy(mathematics, science, engineering)](https://www.scipy.org/). Thus, you can use many of the features included in those packages without importing them at all.

<p><a id="seriesanddataFrame"> </a></p>
<h3 id="seriesanddataFrame">2.1 <a href="#table">Series and DataFrame</a><a class="anchor-link" href="#seriesanddataFrame">¶</a></h3>

The primary objects (Data types) in pandas are:

- ***Series***: a single column of data (one dimension). You can think of *Series* objects as fancier versions of *list*.


- ***DataFrame***: tabular (as in table) data consisting of a collection of Series in which each column is a Pandas series. It is the central data structure used in most analysis using the Pandas library.


- Series can be created through list using **pd.Series(list)**.

In [28]:
# series can be created through list or dictionary
# using list:
city_name = pd.Series(["New York", "Los Angeles", "Chicago"])
city_name

0       New York
1    Los Angeles
2        Chicago
dtype: object

<p><a id="importandexportdata"> </a></p>
<h3 id="importandexportdata">2.2 <a href="#table">Import and export data</a><a class="anchor-link" href="#importandexportdata">¶</a></h3>

- Pandas can read and write dataset in many data formats. To check whether pandas can read your data format, click [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html). 


Let us take an example using the 2012 Beijing housing resale dataset. The dataset was collected from a private housing agency (Lianjia) and then was subsampled to 5,000 transaction records.  

Download the `HouseBeijing2012.xlsx` in today's folder. We are first going to read this Excel as DataFrame into Pandas.
   - function to use: `df = pd.read_excel('<file Path>/HouseBeijing2012.xlsx')`
   - A **file path** is a string that tells the computer where to find a file. 
   - If you have saved `HouseBeijing2012.xlsx` in the same folder as this Jupyter notebook, then all you will need to input as your path is `HouseBeijing2012.xlsx`. 

- To get where you create this Jupyter notebook, we can use the **os** package and the **os.getcwd()** to obtain the working path of the current Jupyter notebook.     

In [7]:
# get the current file path using os.getcwd function
import os
os.getcwd()

'C:\\Users\\Lwz12\\OneDrive\\桌面\\Teaching_CRP4680\\Week2'

In [8]:
# a relative file path
df_2012 = pd.read_excel("HouseBeijing2012.xlsx")
df_2012

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0     BJFT84326414          1544  1400010.56       2012        2           1   
1     BJCP84958845          2606  1800066.00       2012        3           2   
2     BJDX84905788          2264  1350038.34       2012        2           1   
3     BJFT00386624          3621  1800006.91       2012        2           1   
4     BJCY84713854          1127  1970019.58       2012        1           1   
...             ...          ...         ...        ...      ...         ...   
4995  BJTJ84718789          3360  1030038.24       2012        2           1   
4996  BJFT84287006           736  1300028.16       2012        1           1   
4997  BJDC84781079           200  2190056.96       2012        2           1   
4998  BJSJ85075781          2133  1500007.54       2012        1           1   
4999  BJCY84949599           430  1840034.60       2012        2           1   

      Bathroom    Size    FloorLevel  WinSouth  ...  \
0            1   69.68  bottom floor         0  ...   
1            2  129.00           NaN         0  ...   
2            1   88.83             3         1  ...   
3            1   98.69             3         0  ...   
4            1   53.66             4         0  ...   
...        ...     ...           ...       ...  ...   
4995         1   81.84  bottom floor         1  ...   
4996         1   58.12             3         1  ...   
4997         1   58.24             3         1  ...   
4998         1   63.86             2         1  ...   
4999         1   55.19     top floor         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
0     https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
1     https://bj.lianjia.com/xiaoqu/1111027380050/           0             0   
2     https://bj.lianjia.com/xiaoqu/1111027379274/           0             1   
3     https://bj.lianjia.com/xiaoqu/1111027382765/           0             1   
4     https://bj.lianjia.com/xiaoqu/1111027376538/           0             3   
...                                            ...         ...           ...   
4995  https://bj.lianjia.com/xiaoqu/1111027382053/           0             0   
4996  https://bj.lianjia.com/xiaoqu/1111027375508/           0             1   
4997  https://bj.lianjia.com/xiaoqu/1111027374239/           2             1   
4998  https://bj.lianjia.com/xiaoqu/1111027378940/           0             1   
4999  https://bj.lianjia.com/xiaoqu/1111027374756/           0             2   

      Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
0       633.24007            9   1803.02071           0           1   
1      2284.09390            9   9154.80958           0           0   
2       667.21572            8  11158.05983           0           4   
3       939.29061            9   1698.79101           0          10   
4       476.28267            9    938.35742           2           0   
...           ...          ...          ...         ...         ...   
4995   1669.04965            9    948.21718           1           0   
4996    770.89444            9    777.47225           1           1   
4997    715.50723            9   2053.92372           0           0   
4998    863.10517            9    413.21398           2           2   
4999    543.00065            7    522.73335           1           2   

         Dist2CBD  Dist2Center  
0      9345.20091   7396.31505  
1     18298.50637  18632.22305  
2     22480.82065  20105.06770  
3     16309.85203  11427.48851  
4      8105.90581   7213.87518  
...           ...          ...  
4995  15849.26185  21363.35490  
4996   6221.44255   6278.89885  
4997   3779.07141   3505.28158  
4998  21094.05154  15576.85068  
4999   6672.24055   7184.44822  

[5000 rows x 30 columns]

Here, `HouseBeijing2012.xlsx` is our file path. 

- The file path above is a **relative path**, meaning it is relative to the location of this notebook. If you move this notebook to a different folder, the relative path will no longer work. 

- An **absolute path** is a path that is not relative to the location of the notebook. An absolute path will work no matter where the notebook is located

In [32]:
# path of the course folder
path = 'C:\\Users\\Lwz12\\OneDrive\\桌面\\Teaching_CRP4680\\Week2'

In [35]:
# an absolute file path
df_2012 = pd.read_excel(path + "\\HouseBeijing2012.xlsx")
df_2012

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0     BJFT84326414          1544  1400010.56       2012        2           1   
1     BJCP84958845          2606  1800066.00       2012        3           2   
2     BJDX84905788          2264  1350038.34       2012        2           1   
3     BJFT00386624          3621  1800006.91       2012        2           1   
4     BJCY84713854          1127  1970019.58       2012        1           1   
...             ...          ...         ...        ...      ...         ...   
4995  BJTJ84718789          3360  1030038.24       2012        2           1   
4996  BJFT84287006           736  1300028.16       2012        1           1   
4997  BJDC84781079           200  2190056.96       2012        2           1   
4998  BJSJ85075781          2133  1500007.54       2012        1           1   
4999  BJCY84949599           430  1840034.60       2012        2           1   

      Bathroom    Size    FloorLevel  WinSouth  ...  \
0            1   69.68  bottom floor         0  ...   
1            2  129.00           NaN         0  ...   
2            1   88.83             3         1  ...   
3            1   98.69             3         0  ...   
4            1   53.66             4         0  ...   
...        ...     ...           ...       ...  ...   
4995         1   81.84  bottom floor         1  ...   
4996         1   58.12             3         1  ...   
4997         1   58.24             3         1  ...   
4998         1   63.86             2         1  ...   
4999         1   55.19     top floor         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
0     https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
1     https://bj.lianjia.com/xiaoqu/1111027380050/           0             0   
2     https://bj.lianjia.com/xiaoqu/1111027379274/           0             1   
3     https://bj.lianjia.com/xiaoqu/1111027382765/           0             1   
4     https://bj.lianjia.com/xiaoqu/1111027376538/           0             3   
...                                            ...         ...           ...   
4995  https://bj.lianjia.com/xiaoqu/1111027382053/           0             0   
4996  https://bj.lianjia.com/xiaoqu/1111027375508/           0             1   
4997  https://bj.lianjia.com/xiaoqu/1111027374239/           2             1   
4998  https://bj.lianjia.com/xiaoqu/1111027378940/           0             1   
4999  https://bj.lianjia.com/xiaoqu/1111027374756/           0             2   

      Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
0       633.24007            9   1803.02071           0           1   
1      2284.09390            9   9154.80958           0           0   
2       667.21572            8  11158.05983           0           4   
3       939.29061            9   1698.79101           0          10   
4       476.28267            9    938.35742           2           0   
...           ...          ...          ...         ...         ...   
4995   1669.04965            9    948.21718           1           0   
4996    770.89444            9    777.47225           1           1   
4997    715.50723            9   2053.92372           0           0   
4998    863.10517            9    413.21398           2           2   
4999    543.00065            7    522.73335           1           2   

         Dist2CBD  Dist2Center  
0      9345.20091   7396.31505  
1     18298.50637  18632.22305  
2     22480.82065  20105.06770  
3     16309.85203  11427.48851  
4      8105.90581   7213.87518  
...           ...          ...  
4995  15849.26185  21363.35490  
4996   6221.44255   6278.89885  
4997   3779.07141   3505.28158  
4998  21094.05154  15576.85068  
4999   6672.24055   7184.44822  

[5000 rows x 30 columns]

Note: we cannot use backslashes (`\`) alone to construct file path because 
backslashes (`\`) are treated as escape characters in Python strings

- Three Ways to Import a File in Python
1. Use a raw string by adding an `r` in front of the file path:
   ```python
   df = pd.read_csv(r"C:\Users\Documents\data.csv")
    ```
2. Replace backslashes (\) with double backslashes (\\):
    ```python
    df = pd.read_csv("C:\\Users\\Documents\\data.csv")
    ```
3. Replace backslashes (\) with forward slashes (/):
    ```python
    df = pd.read_csv("C:/Users/Documents/data.csv")
    ```

In [12]:
# examples: escape character
print("hello \n \n world! ")

hello 
 
 world! 



Similarly, to export a DataFrame as an Excel: 
   - use function `df.to_excel('<file path>/data.xlsx')`&mdash;export a DataFrame called df to an excel file and save in the designated folder.  

<p><a id="exploringourdata"> </a></p>
<h3 id="exploringourdata">2.3 <a href="#table">Exploring our data</a><a class="anchor-link" href="#exploring our data">¶</a></h3>

You can see here that (in addition to the formatting of tabular data in Jupyter) the main difference between this DataFrame and a series is that we have multiple columns with column labels. So the dataframe structure consists of: 

- An index, with index labels (here the labels are just `0`,`1`,...,`4999`)
- Columns, with column labels (e.g., `HouseID`, `CommunityID`,etc.)
- And the data, which are the values in each row. 

Now let us to some exploration

- To check the top five rows, use **`df.head(5)`**

In [15]:
# check the top five rows: 
df_2012.head(5)

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0  BJFT84326414          1544  1400010.56       2012        2           1   
1  BJCP84958845          2606  1800066.00       2012        3           2   
2  BJDX84905788          2264  1350038.34       2012        2           1   
3  BJFT00386624          3621  1800006.91       2012        2           1   
4  BJCY84713854          1127  1970019.58       2012        1           1   

   Bathroom    Size    FloorLevel  WinSouth  ...  \
0         1   69.68  bottom floor         0  ...   
1         2  129.00           NaN         0  ...   
2         1   88.83             3         1  ...   
3         1   98.69             3         0  ...   
4         1   53.66             4         0  ...   

                                      XIAOQUWEB  SchQuality  NumSubway1km  \
0  https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
1  https://bj.lianjia.com/xiaoqu/1111027380050/           0             0   
2  https://bj.lianjia.com/xiaoqu/1111027379274/           0             1   
3  https://bj.lianjia.com/xiaoqu/1111027382765/           0             1   
4  https://bj.lianjia.com/xiaoqu/1111027376538/           0             3   

   Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m     Dist2CBD  \
0    633.24007            9   1803.02071           0           1   9345.20091   
1   2284.09390            9   9154.80958           0           0  18298.50637   
2    667.21572            8  11158.05983           0           4  22480.82065   
3    939.29061            9   1698.79101           0          10  16309.85203   
4    476.28267            9    938.35742           2           0   8105.90581   

   Dist2Center  
0   7396.31505  
1  18632.22305  
2  20105.06770  
3  11427.48851  
4   7213.87518  

[5 rows x 30 columns]

In [16]:
# check the last five rows
df_2012.tail(5)

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
4995  BJTJ84718789          3360  1030038.24       2012        2           1   
4996  BJFT84287006           736  1300028.16       2012        1           1   
4997  BJDC84781079           200  2190056.96       2012        2           1   
4998  BJSJ85075781          2133  1500007.54       2012        1           1   
4999  BJCY84949599           430  1840034.60       2012        2           1   

      Bathroom   Size    FloorLevel  WinSouth  ...  \
4995         1  81.84  bottom floor         1  ...   
4996         1  58.12             3         1  ...   
4997         1  58.24             3         1  ...   
4998         1  63.86             2         1  ...   
4999         1  55.19     top floor         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
4995  https://bj.lianjia.com/xiaoqu/1111027382053/           0             0   
4996  https://bj.lianjia.com/xiaoqu/1111027375508/           0             1   
4997  https://bj.lianjia.com/xiaoqu/1111027374239/           2             1   
4998  https://bj.lianjia.com/xiaoqu/1111027378940/           0             1   
4999  https://bj.lianjia.com/xiaoqu/1111027374756/           0             2   

      Dist2Subway  HospQuality   Dist2Hosp  NumHosp1km  NumBus200m  \
4995   1669.04965            9   948.21718           1           0   
4996    770.89444            9   777.47225           1           1   
4997    715.50723            9  2053.92372           0           0   
4998    863.10517            9   413.21398           2           2   
4999    543.00065            7   522.73335           1           2   

         Dist2CBD  Dist2Center  
4995  15849.26185  21363.35490  
4996   6221.44255   6278.89885  
4997   3779.07141   3505.28158  
4998  21094.05154  15576.85068  
4999   6672.24055   7184.44822  

[5 rows x 30 columns]

- To explore the shape (dimension) of the DataFrame, use **`df.shape`**

In [17]:
df_2012.shape

(5000, 30)

- To explore the column names, use **`df.columns`**. Note that the `df.columns` returns an object rather than a list. 

In [20]:
# check column names
df_2012.columns

Index(['HouseID', 'CommunityID', 'TotalPrice', 'TransYear', 'Bedroom',
       'Livingroom', 'Bathroom', 'Size', 'FloorLevel', 'WinSouth',
       'WinSouthNorth', 'Decoration', 'TotalFloor', 'BuiltYear', 'Elevation',
       'Heating', 'TransMonth', 'TransDay', 'District', 'CensusTract',
       'XIAOQUWEB', 'SchQuality', 'NumSubway1km', 'Dist2Subway', 'HospQuality',
       'Dist2Hosp', 'NumHosp1km', 'NumBus200m', 'Dist2CBD', 'Dist2Center'],
      dtype='object')

In [21]:
# Get the index of the variable "Bedroom"
list(df_2012.columns).index("Bedroom")

4

- Basic descriptive statistics for our dataframe **df.describe()**

In [23]:
# descriptive statistics
df_2012.describe()

CommunityID    TotalPrice  TransYear      Bedroom   Livingroom  \
count  5000.000000  5.000000e+03     5000.0  5000.000000  5000.000000   
mean   1896.095000  1.947660e+06     2012.0     1.928600     1.207600   
std    1033.137016  9.321690e+05        0.0     0.694837     0.505721   
min       4.000000  4.290000e+05     2012.0     1.000000     0.000000   
25%     981.750000  1.325027e+06     2012.0     1.000000     1.000000   
50%    1968.500000  1.750016e+06     2012.0     2.000000     1.000000   
75%    2752.000000  2.300078e+06     2012.0     2.000000     2.000000   
max    3718.000000  1.330012e+07     2012.0     4.000000     2.000000   

          Bathroom         Size     WinSouth  WinSouthNorth  Decoration  ...  \
count  5000.000000  5000.000000  5000.000000    5000.000000      5000.0  ...   
mean      1.152600    80.306600     0.764400       0.468200         0.0  ...   
std       0.367342    29.529724     0.424416       0.499038         0.0  ...   
min       1.000000    24.400000     0.000000       0.000000         0.0  ...   
25%       1.000000    58.457500     1.000000       0.000000         0.0  ...   
50%       1.000000    73.390000     1.000000       0.000000         0.0  ...   
75%       1.000000    95.852500     1.000000       1.000000         0.0  ...   
max       3.000000   292.930000     1.000000       1.000000         0.0  ...   

       CensusTract   SchQuality  NumSubway1km  Dist2Subway  HospQuality  \
count  5000.000000  5000.000000   5000.000000  5000.000000  5000.000000   
mean    119.434200     0.433800      0.804400  1183.742821     8.634400   
std      63.883396     1.020895      0.834674  1036.077987     0.684127   
min       2.000000     0.000000      0.000000    42.986150     7.000000   
25%      64.000000     0.000000      0.000000   567.024090     9.000000   
50%     118.000000     0.000000      1.000000   867.046620     9.000000   
75%     177.000000     0.000000      1.000000  1396.076155     9.000000   
max     226.000000     4.000000      4.000000  8107.925470     9.000000   

          Dist2Hosp   NumHosp1km   NumBus200m      Dist2CBD   Dist2Center  
count   5000.000000  5000.000000  5000.000000   5000.000000   5000.000000  
mean    3202.480164     0.245800     3.146400  13449.129254  13063.810653  
std     3374.224481     0.541702     4.734242   6524.513313   6682.127691  
min       53.190330     0.000000     0.000000    990.319030    968.061030  
25%     1164.416550     0.000000     0.000000   8827.193230   7543.081360  
50%     1995.616470     0.000000     1.000000  12725.224445  11571.991570  
75%     3799.902080     0.000000     5.000000  18418.616755  17531.304565  
max    23348.074690     4.000000    42.000000  37268.650970  40228.303480  

[8 rows x 27 columns]

- `.sort_values()` sorts your DataFrame by a certain column. If you column is numeric, it will sort the values from smallest to largest. If your column is a string, it will sort alphabetically.
    - The index values will be sorted along with the rows. This means not only are the row values rearranged, but their corresponding index values move along with them - the index stays associated with the same row,
    - By default, `df.sort_values()` does not modify the DataFrame in place. Instead, it returns a new DataFrame with the values sorted. To apply the sorting directly to the original DataFrame,  use the `inplace=True` parameter.
      - inplace=True: Modifies the original DataFrame directly without returning a new DataFrame.
      - inplace=False: Returns a new sorted DataFrame, leaving the original DataFrame unchanged.

In [26]:
# sort from the smallest to the largest
df_2012.sort_values("TotalPrice")

HouseID  CommunityID   TotalPrice  TransYear  Bedroom  Livingroom  \
4572  BJHD84586553          2268    429000.00       2012        2           1   
4360  BJDX84234421          1792    465041.19       2012        1           2   
668   BJSY84820479          3065    500024.60       2012        1           0   
1832  BJCY84803356          2215    520032.42       2012        1           1   
4833  BJCY84404564          2215    550017.70       2012        1           1   
...             ...          ...          ...        ...      ...         ...   
886   BJXC84867016          3206   7916136.18       2012        4           1   
1437  BJHD84925549          3518   8250036.00       2012        3           2   
738   BJHD84822957           376   8540189.46       2012        3           2   
1113  BJCY84196120          3470   9500207.97       2012        4           2   
3645  BJHD84357872          2732  13300123.74       2012        4           2   

      Bathroom    Size FloorLevel  WinSouth  ...  \
4572         1   52.80          3         1  ...   
4360         1   67.29          3         1  ...   
668          1   32.66          3         1  ...   
1832         1   36.78  top floor         1  ...   
4833         1   33.70  top floor         1  ...   
...        ...     ...        ...       ...  ...   
886          2  147.34          3         1  ...   
1437         2  157.00          2         1  ...   
738          2  190.01          3         1  ...   
1113         3  229.79          2         1  ...   
3645         3  249.22          3         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
4572  https://bj.lianjia.com/xiaoqu/1111027379278/           1             2   
4360  https://bj.lianjia.com/xiaoqu/1111027378083/           0             0   
668   https://bj.lianjia.com/xiaoqu/1111027381324/           0             1   
1832  https://bj.lianjia.com/xiaoqu/1111027379159/           0             0   
4833  https://bj.lianjia.com/xiaoqu/1111027379159/           0             0   
...                                            ...         ...           ...   
886   https://bj.lianjia.com/xiaoqu/1111027381668/           4             1   
1437  https://bj.lianjia.com/xiaoqu/1111027382487/           2             0   
738   https://bj.lianjia.com/xiaoqu/1111027374627/           2             1   
1113  https://bj.lianjia.com/xiaoqu/1111027382344/           0             1   
3645  https://bj.lianjia.com/xiaoqu/1111027380422/           2             2   

      Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
4572    494.93863            9   1018.48756           0           0   
4360   5480.12539            8  22013.62655           0           1   
668     160.09919            9  15515.11968           0          11   
1832   1712.15277            9   6616.04165           0           2   
4833   1712.15277            9   6616.04165           0           2   
...           ...          ...          ...         ...         ...   
886     888.84356            9   1375.33226           0           1   
1437   1297.85840            9    896.61896           1           3   
738     495.85158            9   1337.76996           0           0   
1113    331.75223            7   1546.11645           0           0   
3645    498.20211            9   2533.70859           0           7   

         Dist2CBD  Dist2Center  
4572  12814.51532   7597.30985  
4360  34020.54729  32407.04761  
668   29219.05096  32530.22463  
1832  19522.51032  23249.49135  
4833  19522.51032  23249.49135  
...           ...          ...  
886    8903.54161   6298.83422  
1437  15310.60488   9820.69326  
738   15068.05800   9903.45366  
1113   6198.60201   7944.01154  
3645  15680.12935  11144.58343  

[5000 rows x 30 columns]

In [27]:
# sort from the largest to the smallest
df_2012.sort_values("TotalPrice", ascending = False)

HouseID  CommunityID   TotalPrice  TransYear  Bedroom  Livingroom  \
3645  BJHD84357872          2732  13300123.74       2012        4           2   
1113  BJCY84196120          3470   9500207.97       2012        4           2   
738   BJHD84822957           376   8540189.46       2012        3           2   
1437  BJHD84925549          3518   8250036.00       2012        3           2   
886   BJXC84867016          3206   7916136.18       2012        4           1   
...             ...          ...          ...        ...      ...         ...   
4833  BJCY84404564          2215    550017.70       2012        1           1   
1832  BJCY84803356          2215    520032.42       2012        1           1   
668   BJSY84820479          3065    500024.60       2012        1           0   
4360  BJDX84234421          1792    465041.19       2012        1           2   
4572  BJHD84586553          2268    429000.00       2012        2           1   

      Bathroom    Size FloorLevel  WinSouth  ...  \
3645         3  249.22          3         1  ...   
1113         3  229.79          2         1  ...   
738          2  190.01          3         1  ...   
1437         2  157.00          2         1  ...   
886          2  147.34          3         1  ...   
...        ...     ...        ...       ...  ...   
4833         1   33.70  top floor         1  ...   
1832         1   36.78  top floor         1  ...   
668          1   32.66          3         1  ...   
4360         1   67.29          3         1  ...   
4572         1   52.80          3         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
3645  https://bj.lianjia.com/xiaoqu/1111027380422/           2             2   
1113  https://bj.lianjia.com/xiaoqu/1111027382344/           0             1   
738   https://bj.lianjia.com/xiaoqu/1111027374627/           2             1   
1437  https://bj.lianjia.com/xiaoqu/1111027382487/           2             0   
886   https://bj.lianjia.com/xiaoqu/1111027381668/           4             1   
...                                            ...         ...           ...   
4833  https://bj.lianjia.com/xiaoqu/1111027379159/           0             0   
1832  https://bj.lianjia.com/xiaoqu/1111027379159/           0             0   
668   https://bj.lianjia.com/xiaoqu/1111027381324/           0             1   
4360  https://bj.lianjia.com/xiaoqu/1111027378083/           0             0   
4572  https://bj.lianjia.com/xiaoqu/1111027379278/           1             2   

      Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
3645    498.20211            9   2533.70859           0           7   
1113    331.75223            7   1546.11645           0           0   
738     495.85158            9   1337.76996           0           0   
1437   1297.85840            9    896.61896           1           3   
886     888.84356            9   1375.33226           0           1   
...           ...          ...          ...         ...         ...   
4833   1712.15277            9   6616.04165           0           2   
1832   1712.15277            9   6616.04165           0           2   
668     160.09919            9  15515.11968           0          11   
4360   5480.12539            8  22013.62655           0           1   
4572    494.93863            9   1018.48756           0           0   

         Dist2CBD  Dist2Center  
3645  15680.12935  11144.58343  
1113   6198.60201   7944.01154  
738   15068.05800   9903.45366  
1437  15310.60488   9820.69326  
886    8903.54161   6298.83422  
...           ...          ...  
4833  19522.51032  23249.49135  
1832  19522.51032  23249.49135  
668   29219.05096  32530.22463  
4360  34020.54729  32407.04761  
4572  12814.51532   7597.30985  

[5000 rows x 30 columns]

<p><a id="indexingandslicingofdataFrame"> </a></p>
<h3 id="indexingandslicingofdataFrame">2.4 <a href="#table">Indexing and slicing DataFrame</a><a class="anchor-link" href="#indexingandslicingofdataFrame">¶</a></h3>

selecting a subset of a DataFrame using **Indexing** and **Slicing**. 
- *Indexing* means selecting particular row or column from a DataFrame. 
- *Slicing* means selecting multiple rows and columns

Three ways of selecting particular rows and columns of an DataFrame object&mdash; `df[]`, `df.loc[rows_label , columns_label]` and `df.iloc[row_position , column_position]`. 
- **label** and **position** 
- A *label*: one name in the column list or an index in the row index (the column at far left). 
- A *position*: the corresponding position of column name or index in a sequence, starting from zero.

**1. Selecting elements based on `df[]`**: 
- `df["col_name_1"]` select the column named "col_name1", and return a Series.
- `df[["col_name_1", "col_name_5"]]`, select multiple columns together, and return a DataFrame.

In [33]:
# select HouseID based on column: 
df_2012["HouseID"]


0       BJFT84326414 
1       BJCP84958845 
2       BJDX84905788 
3       BJFT00386624 
4       BJCY84713854 
            ...      
4995    BJTJ84718789 
4996    BJFT84287006 
4997    BJDC84781079 
4998    BJSJ85075781 
4999    BJCY84949599 
Name: HouseID, Length: 5000, dtype: object


- try type in `df[["col_name_1"]]`, how it differs from `df["col_name_1"]`?


In [35]:
type(df_2012[["HouseID"]])

pandas.core.frame.DataFrame

In [36]:
# select multiple columns, all rows
df_2012[["HouseID", "TotalPrice"]]

HouseID  TotalPrice
0     BJFT84326414   1400010.56
1     BJCP84958845   1800066.00
2     BJDX84905788   1350038.34
3     BJFT00386624   1800006.91
4     BJCY84713854   1970019.58
...             ...         ...
4995  BJTJ84718789   1030038.24
4996  BJFT84287006   1300028.16
4997  BJDC84781079   2190056.96
4998  BJSJ85075781   1500007.54
4999  BJCY84949599   1840034.60

[5000 rows x 2 columns]

**2. Selecting elements based on `df.loc[ rows_label , columns_label ]`**:


In [42]:
# select the first 2 rows and all columns
df_2012.loc[ 0:1  ,   ]

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0  BJFT84326414          1544  1400010.56       2012        2           1   
1  BJCP84958845          2606  1800066.00       2012        3           2   

   Bathroom    Size    FloorLevel  WinSouth  ...  \
0         1   69.68  bottom floor         0  ...   
1         2  129.00           NaN         0  ...   

                                      XIAOQUWEB  SchQuality  NumSubway1km  \
0  https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
1  https://bj.lianjia.com/xiaoqu/1111027380050/           0             0   

   Dist2Subway  HospQuality   Dist2Hosp  NumHosp1km  NumBus200m     Dist2CBD  \
0    633.24007            9  1803.02071           0           1   9345.20091   
1   2284.09390            9  9154.80958           0           0  18298.50637   

   Dist2Center  
0   7396.31505  
1  18632.22305  

[2 rows x 30 columns]

In [47]:
# select the first 3 rows and column "CommunityID" on a dataframe, and returns a dataframe
df_2012.loc[ 0:2    ,  ["CommunityID"]  ]

CommunityID
0         1544
1         2606
2         2264

In [49]:
# select the first 3 rows, and the columns "col_name_m" and "col_name_n"
df_2012.loc[  :2 ,   ["CommunityID", "TotalPrice"]]

CommunityID  TotalPrice
0         1544  1400010.56
1         2606  1800066.00
2         2264  1350038.34

In [59]:
# select the rows with label 3 and 5, and columns from "HouseID" to "TotalPrice"
df_2012.loc[ [3, 5] , [ "HouseID", "TotalPrice" ]  ]

HouseID  TotalPrice
3  BJFT00386624   1800006.91
5  BJFT85228189   1280012.60

**3. Selecting elements based on `df.iloc[ row_position , column_position ]`:**

In [52]:
# select first 3 rows and the second column
df_2012.iloc[ 0:3  , [1]  ]

CommunityID
0         1544
1         2606
2         2264

In [55]:
# select first 3 rows and columns "HouseID", "TotalPrice": list().index()
df_2012.iloc[ 0:3  ,  [list(df_2012.columns).index("HouseID") , list(df_2012.columns).index("TotalPrice") ]  ]

HouseID  TotalPrice
0  BJFT84326414   1400010.56
1  BJCP84958845   1800066.00
2  BJDX84905788   1350038.34

In [61]:
# select all rows and the last 2 columns "Dist2CBD" and "Dist2Center"
df_2012.iloc[ :  ,  -2:   ]

Dist2CBD  Dist2Center
0      9345.20091   7396.31505
1     18298.50637  18632.22305
2     22480.82065  20105.06770
3     16309.85203  11427.48851
4      8105.90581   7213.87518
...           ...          ...
4995  15849.26185  21363.35490
4996   6221.44255   6278.89885
4997   3779.07141   3505.28158
4998  21094.05154  15576.85068
4999   6672.24055   7184.44822

[5000 rows x 2 columns]

<p><a id="filtering"> </a></p>
<h3 id="filtering">2.5 <a href="#table">Filtering</a><a class="anchor-link" href="#filtering">¶</a></h3>

Let us filter the DataFrame based on specified conditions. Filtering in pandas is done with a boolean expression. The expression is evaluated for each row in the dataframe, and only rows where the expression evaluates to True are returned.

Let us say we want to select those housing samples located within 1500m of the subway stations. Those units are usually marked as "subway housing" for which buyers are willing to pay more.

In [68]:
# select housing samples located within 1500m of the subway stations
df_subway = df_2012[df_2012["Dist2Subway"] <=1500]
df_subway

# You can always leave out the column argument in either .loc or .iloc in order to select all columns.

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0     BJFT84326414          1544  1400010.56       2012        2           1   
2     BJDX84905788          2264  1350038.34       2012        2           1   
3     BJFT00386624          3621  1800006.91       2012        2           1   
4     BJCY84713854          1127  1970019.58       2012        1           1   
6     BJCY84112518          2767  2730093.60       2012        3           2   
...             ...          ...         ...        ...      ...         ...   
4994  BJCY84422499           109  1940059.80       2012        2           1   
4996  BJFT84287006           736  1300028.16       2012        1           1   
4997  BJDC84781079           200  2190056.96       2012        2           1   
4998  BJSJ85075781          2133  1500007.54       2012        1           1   
4999  BJCY84949599           430  1840034.60       2012        2           1   

      Bathroom    Size    FloorLevel  WinSouth  ...  \
0            1   69.68  bottom floor         0  ...   
2            1   88.83             3         1  ...   
3            1   98.69             3         0  ...   
4            1   53.66             4         0  ...   
6            2  132.08             2         1  ...   
...        ...     ...           ...       ...  ...   
4994         1   67.41             3         1  ...   
4996         1   58.12             3         1  ...   
4997         1   58.24             3         1  ...   
4998         1   63.86             2         1  ...   
4999         1   55.19     top floor         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
0     https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
2     https://bj.lianjia.com/xiaoqu/1111027379274/           0             1   
3     https://bj.lianjia.com/xiaoqu/1111027382765/           0             1   
4     https://bj.lianjia.com/xiaoqu/1111027376538/           0             3   
6     https://bj.lianjia.com/xiaoqu/1111027380518/           0             1   
...                                            ...         ...           ...   
4994  https://bj.lianjia.com/xiaoqu/1111027374001/           0             0   
4996  https://bj.lianjia.com/xiaoqu/1111027375508/           0             1   
4997  https://bj.lianjia.com/xiaoqu/1111027374239/           2             1   
4998  https://bj.lianjia.com/xiaoqu/1111027378940/           0             1   
4999  https://bj.lianjia.com/xiaoqu/1111027374756/           0             2   

      Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
0       633.24007            9   1803.02071           0           1   
2       667.21572            8  11158.05983           0           4   
3       939.29061            9   1698.79101           0          10   
4       476.28267            9    938.35742           2           0   
6       692.53062            9   5122.56097           0           0   
...           ...          ...          ...         ...         ...   
4994   1083.57589            9   1131.66080           0          30   
4996    770.89444            9    777.47225           1           1   
4997    715.50723            9   2053.92372           0           0   
4998    863.10517            9    413.21398           2           2   
4999    543.00065            7    522.73335           1           2   

         Dist2CBD  Dist2Center  
0      9345.20091   7396.31505  
2     22480.82065  20105.06770  
3     16309.85203  11427.48851  
4      8105.90581   7213.87518  
6     11524.68492  16967.87510  
...           ...          ...  
4994   9079.72181  10821.04163  
4996   6221.44255   6278.89885  
4997   3779.07141   3505.28158  
4998  21094.05154  15576.85068  
4999   6672.24055   7184.44822  

[3919 rows x 30 columns]

- `df.loc[df["Dist2Subway"] <= 1500, : ]`:
  - step1, `df["Dist2Subway"] <= 1500` return a series with values of ***False*** or ***True*** (boolean type); this is known as **boolean indexing**
  - step2, it is enclosed by `df.loc[]` and can return a subset of the candidate rows
  - step3, assign the returned DataFrame to a new dataframe called `df_subway`

In order to filter by more than one condition, you must:
- Put all conditions in `()`
- Separate the condtions by: 
    - `|` if an `OR` condition     
    - `&` if an `AND` condition

In [84]:
# houses within 1500m of subway stations with at least two Bathrooms

df_2012[(df_2012["Dist2Subway"] <= 1500) & (df_2012["Bathroom"] >= 2)]

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
6     BJCY84112518          2767  2730093.60       2012        3           2   
16    BJHD84735168            73  4800000.00       2012        3           2   
20    BJCP84427284          2624  1945008.00       2012        3           1   
24    BJCY84206540           219  5690099.85       2012        3           2   
26    BJDX84708873          3293  2350086.96       2012        3           2   
...             ...          ...         ...        ...      ...         ...   
4947  BJCY84980409           428  4550118.30       2012        2           2   
4964  BJSY84514643          2092  1440096.45       2012        3           2   
4965  BJCY84712249          1023  2270089.20       2012        2           1   
4970  BJTJ00464885          1222  1690015.62       2012        3           2   
4977  BJDC84682352          2594  3880128.40       2012        2           2   

      Bathroom    Size FloorLevel  WinSouth  ...  \
6            2  132.08          2         1  ...   
16           2  162.31          2         1  ...   
20           2  124.68  top floor         1  ...   
24           2  132.59          4         1  ...   
26           2  131.82          3         1  ...   
...        ...     ...        ...       ...  ...   
4947         2  123.46          3         1  ...   
4964         2  116.09  top floor         1  ...   
4965         2   97.95          4         1  ...   
4970         2  124.66          3         1  ...   
4977         2  131.05          4         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
6     https://bj.lianjia.com/xiaoqu/1111027380518/           0             1   
16    https://bj.lianjia.com/xiaoqu/1111027373903/           1             2   
20    https://bj.lianjia.com/xiaoqu/1111027380088/           0             0   
24    https://bj.lianjia.com/xiaoqu/1111027374273/           3             1   
26    https://bj.lianjia.com/xiaoqu/1111027381865/           0             0   
...                                            ...         ...           ...   
4947  https://bj.lianjia.com/xiaoqu/1111027374751/           0             1   
4964  https://bj.lianjia.com/xiaoqu/1111027378834/           0             1   
4965  https://bj.lianjia.com/xiaoqu/1111027376256/           0             1   
4970  https://bj.lianjia.com/xiaoqu/1111027376743/           0             2   
4977  https://bj.lianjia.com/xiaoqu/1111027380027/           0             0   

      Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
6       692.53062            9   5122.56097           0           0   
16      780.62340            9   2388.29366           0           2   
20     1353.49228            9   2995.00650           0           4   
24      584.59806            9   2425.10859           0           0   
26     1183.40088            8   2576.24539           0           5   
...           ...          ...          ...         ...         ...   
4947    133.56460            9    901.77684           2           0   
4964    720.06897            9  15070.14248           0           0   
4965    730.91213            9   6431.98463           0           1   
4970    584.72264            9   3799.90208           0           1   
4977   1453.55076            9   1403.04328           0           0   

         Dist2CBD  Dist2Center  
6     11524.68492  16967.87510  
16    12931.55797   8841.02198  
20    22156.56301  20492.28868  
24     1150.59048   5314.35936  
26    11878.33320  12565.01538  
...           ...          ...  
4947   7059.91186   7359.74676  
4964  28857.17822  32072.73480  
4965  14835.75290  14871.11849  
4970  19441.12723  24831.94635  
4977   6795.08180   4779.29945  

[592 rows x 30 columns]

Note: 
- Here, you cannot use the `.iloc` function becuase it used positional indexing. Boolean indexing (e.g., `(df_2012["Dist2Subway"] <= 1500) & (df_2012["Bathroom"] >= 2))` is used to filter rows or columns based on conditions, and it works with `df.loc[]` or directly within square brackets `df[]`.

<p><a id="modifyingelements"> </a></p>
<h3 id="modifyingelements">2.6 <a href="#table">Modifying elements</a><a class="anchor-link" href="#modifyingelements">¶</a></h3>

Essentially, in the same way that we can select elements we can also update them using the same logic.

In [78]:
# change the total price of a record: 4800155.94 -> 4800000
df_2012.loc[ df_2012["TotalPrice"] == 4800155.94 ,  "TotalPrice" ] = 4800000


Now, let us consider a more comprehensive example. Let us say I am interesting in comparing the average housing prices for the houses that are located at different distances from the nearest subway station.

To achieve this goal, we firstly create the categorical variable, denoting the subway accessibility. We can define a categorical variable 'Subwaylevel' that categorizes the continuous variable "Dist2Subway" (distance to subway) into 3 levels, 0-500m, 500-1500m, and beyond 1500m.

In [99]:
# note: we create a new variable - Sublevel
# It is not recommended to directly modify the original elements. 
df_2012["Sublevel"] = "Level 1"
# (df_2012["Dist2Subway"] > 500) & (df_2012["Dist2Subway"] <= 1500)
df_2012.loc[(df_2012["Dist2Subway"] > 500) & (df_2012["Dist2Subway"] <= 1500) , "Sublevel"] = "Level 2"
df_2012.loc[(df_2012["Dist2Subway"] > 1500) , "Sublevel"] = "Level 3"

<p><a id="pandasarithmetics"> </a></p>
<h3 id="pandasarithmetics">2.7 <a href="#table">Pandas Arithmetics</a><a class="anchor-link" href="#pandasarithmetics">¶</a></h3>


**1. calculate unit housing price** 
We only have "TotalPrice" by now, representing the total amount of money spent for a house. However, the normalized price (unitprice) makes more sense. So let us calculate the unit price

In [89]:
df_2012.columns

Index(['HouseID', 'CommunityID', 'TotalPrice', 'TransYear', 'Bedroom',
       'Livingroom', 'Bathroom', 'Size', 'FloorLevel', 'WinSouth',
       'WinSouthNorth', 'Decoration', 'TotalFloor', 'BuiltYear', 'Elevation',
       'Heating', 'TransMonth', 'TransDay', 'District', 'CensusTract',
       'XIAOQUWEB', 'SchQuality', 'NumSubway1km', 'Dist2Subway', 'HospQuality',
       'Dist2Hosp', 'NumHosp1km', 'NumBus200m', 'Dist2CBD', 'Dist2Center',
       'Sublevel'],
      dtype='object')

In [100]:
# UnitPrice = TotalPrice/Size
df_2012["UnitPrice"] = df_2012["TotalPrice"]/df_2012["Size"]
df_2012.sort_values("UnitPrice", ascending = False).UnitPrice
# Pandas arithmetics are vectorized, meaning that the operation is applied to each element in the column

2437    66845.0
2186    63254.0
1381    61112.0
4944    60247.0
3827    59458.0
         ...   
3599     8393.0
4572     8125.0
865      7856.0
1762     7112.0
4360     6911.0
Name: UnitPrice, Length: 5000, dtype: float64

<p><a id="groupby"> </a></p>
<h3 id="groupby">2.8 <a href="#table">Groupby function</a><a class="anchor-link" href="#groupby">¶</a></h3>


Now we want to group the houses based on `Sublevel`, and calculate the average housing price for each group. 

- `pd.groupby()`
    - similar to the pivotal table in Excel
    - involves three main steps
        - Splitting: Divides the DataFrame into groups based on a column.
        - Applying: Applies a function (e.g., mean, sum) to each group.
        - Combining: Combines the results into a new table.
    - In our case
        - Split the DataFrame by Sublevel (distance categories to subway).
        - Apply a function (e.g., mean()) to calculate the average price.
        - Combine the grouped results to form a summary table.- 


    - Notice that `pd.groupby()` does not return a DataFrame
        - we need to use a function, e.g., sum(), mean(), or apply() to make the return a DataFrame.

In [119]:
# groupby applied to UnitPrice
df_new = df_2012.groupby("Sublevel")[["UnitPrice"]].mean()
df_new

UnitPrice
Sublevel              
Level 1   26882.939484
Level 2   25939.017190
Level 3   20328.317299

`.reset_index()` is a method that resets the index of a dataframe to a column of your choice. The default is to reset the index to a column of sequential numbers.

- `drop`: If True, the current index is removed completely and is not added as a column. The default is False.
- `inplace`: If True, the DataFrame is modified in place, and the method does not return a new DataFrame. The default is False.

In [121]:
# .reset_index()
df_new.reset_index(drop = False, inplace = True)
df_new

Sublevel     UnitPrice
0  Level 1  26882.939484
1  Level 2  25939.017190
2  Level 3  20328.317299

In [133]:
# groupby sublevel and get the average values of "UnitPrice","Dist2Subway", and "Size"
df_new = df_2012.groupby("Sublevel")[["UnitPrice","Dist2Subway", "Size"]].mean().reset_index()
df_new

Sublevel     UnitPrice  Dist2Subway       Size
0  Level 1  26882.939484   355.116527  81.074792
1  Level 2  25939.017190   902.736946  79.046465
2  Level 3  20328.317299  2713.126174  82.983673

## in-class exercise (due Jan 30th)

1. Import and Explore a Dataset - Find a dataset of your interest, read it into a Pandas DataFrame, and explore its structure.
- Display the first few rows
- Print the variable (column) names
- Print the dimensions (number of rows and columns)

## in-class exercise (due Feb 6th)

2. Import the `HouseBeijing2012.xlsx` data.
- Select the last two rows and the last three columns using both `df.loc[]` and `df.iloc[]` methods.
- Select the first three rows and the column "TransYear" and "Dist2Subway" using both `df.loc[]` and `df.iloc[]` methods.

3. `str1 = "12322378910"`, return a list which stores all indices with value "2". 

4. The unit of the `UnitPrice` we calculated is Chinese Dollar per square meters. (1) Convert the measurement unit to USD per sqft. (2) show the UnitPrice in descending order.  

5. Using the `HouseBeijing2012.xlsx` dataset, filter the dataset to find all the houses with:
- (1) A total price greater than 5 million CNY and 
- (2) A distance to the nearest subway station (Dist2Subway) less than or equal to 1000 meters.

Display the filtered dataset 

6. Group the dataset by the number of bedrooms (Bedroom) and calculate:
- The average unit price for each group.